### PR TITLE
Create "light" test-running mode and a '--noLibCheck' flag

### PR DESCRIPTION
**Pull request by DanielRosenwasser**
# Update

*Light mode now only skips invariant checking and parent pointer setup.*

## How to review this PR

Read <a href="#the-description">the description</a> and probably review commit-by-commit.

## <a name="the-description">The Description</a>

This pull request creates a new mode for running tests a little faster, and also modifies the way in which tests are run.

## Light mode

Light mode is a new flag for running tests so that issues that generally don't (*shouldn't?*) pop up can be ignored. This enables devs to iterate a bit more quickly and tightens the feedback loop when making changes. 

Light mode does this by
* avoiding tree invariant checking
* avoiding performing a *full* check phase of `lib.d.ts`, since
    * it should always be correct
    * types are lazily computed (i.e. "helicoptering")
    * duplicate identifier errors can be caught at the binding phase

## The `--noLibCheck` flag

`--noLibCheck` avoids type-checking files that have a `no-default-lib` reference comment. We use this for light mode, since `lib.d.ts` and `lib.es6.d.ts` will always have said reference comment.

## Improvements

Here's what I jsut got for one run of each. Other runs have generally been consistent with the below results.

Mode | Time to Run | Human Readable Time to Run
-------|-----------------|----------------------------
Normal Mode w/  Invariant Checking | 248070 ms | 4 minutes,  8.1 seconds
Normal Mode w/o Invariant Checking | 213312 ms | 3 minutes, 33.3 seconds
Light Mode                         | 159424 ms | 2 minutes, 39.5 seconds

## Testing differences

Things that are now quite different in the test harness are that:
* `lib.d.ts` is always processed *last* to get consistent ordering for union types.
* `lib.d.ts` is now always included unless `// @noLib: true` is included at the top of your file.
    * This is true even if a `no-default-lib` reference comment is encountered.

## Long term plan

As we find other checks we can avoid, light mode can encompass those too.

CI testing will always use the original testing mode, but light mode may become the default and the current mode may become something like "full mode" when we are confident that light mode isn't disruptive and inconsistent.

### CODE REVIEW

File: src/harness/runnerServer.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
+﻿
```
Comment (danquirk) [inline src/harness/runnerServer.ts]: Assuming this is not ready to go in :)

**Comment by danquirk** [comment]
_so that issues that generally don't (shouldn't?) pop up can be ignored_

So do we believe this strongly at the moment? If so do we want to just enable this by default? Or do you want to give it some time to bake and let people manually use it and ensure they're not missing things?

We can give Travis a different test target than runtests where light mode is off but then local dev runs get the perf speedup without needing to knowingly opt in.

File: src/harness/runnerServer.ts (hunk 1 above)
Comment (DanielRosenwasser) [inline src/harness/runnerServer.ts]: Whoops! Yes, this was me prototyping for a different improvement and then forking off (pun intended) to work on this.

**Comment by DanielRosenwasser** [comment]
> Or do you want to give it some time to bake and let people manually use it and ensure they're not missing things?

I think that was the original intent, though, given how few baselines were affected, I can't see what sorts of issues we'd run into with it not being the default. We should see if the general consensus on the team is to make light mode the immediate default, and make a full mode that does *all* the checks. I'd probably be open to this. Are you weighing in this direction personally?

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -10957,7 +10957,14 @@ module ts {
 
         function checkSourceFile(node: SourceFile) {
             let start = new Date().getTime();
-            checkSourceFileWorker(node);
+
+            // Check whether the file has declared it is the default lib,
+            // and whether the user has specifically chosen to avoid checking it.
+            let skipCheck = node.hasNoDefaultLib && compilerOptions.noLibCheck;
```
Comment (CyrusNajmabadi) [inline src/compiler/checker.ts]: hasNoDefaultLib ?

File: src/compiler/program.ts [hunk 3]
```diff
@@ -151,18 +151,25 @@ module ts {
...
-        let seenNoDefaultLib = options.noLib;
         let commonSourceDirectory: string;
         let diagnosticsProducingTypeChecker: TypeChecker;
         let noDiagnosticsTypeChecker: TypeChecker;
 
+        // shouldExcludeDefaultLib is true if:
+        //  - The '--noLib' flag is used.
+        //  - A 'no-default-lib' reference comment is encountered in
+        //      processing the root files.
+        let shouldExcludeDefaultLib = options.noLib;
```
Comment (CyrusNajmabadi) [inline src/compiler/program.ts]: your code doesn't seem to match your comment.

Comment (CyrusNajmabadi) [inline src/compiler/program.ts]: This seems unecessary.  why not have a "let sawNoDefaultLibTag" which you set while processing.  Then , have your check be:

```if (!sawNoDefaultLib && !options.noLib) { ... }```

Merging the two disparate systems into one variable, which you mutate on the fly seems much harder to understand.

File: src/harness/runner.ts [hunk 4]
```diff
@@ -39,40 +40,50 @@ var testConfigFile =
...
-                runners.push(new GeneratedFourslashRunner(FourSlashTestType.Native));
-                break;
-            case 'rwc':
-                runners.push(new RWCRunner());
-                break;
-            case 'test262':
-                runners.push(new Test262BaselineRunner());
-                break;
+    var testConfig = JSON.parse(testConfigFile);
+    //var testConfig = testConfigFile.match(/test:\s\['(.*)'\]/);
```
Comment (CyrusNajmabadi) [inline src/harness/runner.ts]: remove.

File: src/compiler/program.ts (hunk 3 above)
Comment (DanielRosenwasser) [inline src/compiler/program.ts]: True, I was only keeping the core logic the same, but I agree.

File: src/compiler/checker.ts (hunk 2 above)
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: I'm not sure what your question is. :smile:

**Comment by jbondc** [comment]
Oh man, anything to improve tests speed :sparkles: 
It would be nice if not all baselines were deleted.
jake runtests light=true skip=projectOutput

don't delete reference/projectOutput 
don't generate tests in local/projectOutput

File: src/harness/harness.ts [hunk 5]
```diff
@@ -808,10 +809,10 @@ module Harness {
             }
         }
 
-        export function createSourceFileAndAssertInvariants(fileName: string, sourceText: string, languageVersion: ts.ScriptTarget, assertInvariants = true) {
+        export function createSourceFileAndAssertInvariants(fileName: string, sourceText: string, languageVersion: ts.ScriptTarget, lightMode: boolean) {
             // Only set the parent nodes if we're asserting invariants.  We don't need them otherwise.
```
Comment (CyrusNajmabadi) [inline src/harness/harness.ts]: This comment seems wrong now.  also, why did this change?

File: tests/baselines/reference/typeCheckTypeArgument.js [hunk 6]
```diff
@@ -1,5 +1,4 @@
 //// [typeCheckTypeArgument.ts]
-/// <reference no-default-lib="true"/>
```
Comment (CyrusNajmabadi) [inline tests/baselines/reference/typeCheckTypeArgument.js]: Why was this removed?

File: tests/baselines/reference/templateStringsArrayTypeRedefinedInES6Mode.errors.txt [hunk 7]
```diff
@@ -1,10 +1,8 @@
 lib.d.ts(521,11): error TS2300: Duplicate identifier 'TemplateStringsArray'.
 tests/cases/compiler/templateStringsArrayTypeRedefinedInES6Mode.ts(2,7): error TS2300: Duplicate identifier 'TemplateStringsArray'.
-tests/cases/compiler/templateStringsArrayTypeRedefinedInES6Mode.ts(8,3): error TS2345: Argument of type '{ [x: number]: undefined; }' is not assignable to parameter of type 'TemplateStringsArray'.
```
Comment (CyrusNajmabadi) [inline tests/baselines/reference/templateStringsArrayTypeRedefinedInES6Mode.errors.txt]: what happened to this?

File: tests/baselines/reference/typeCheckTypeArgument.js (hunk 6 above)
Comment (DanielRosenwasser) [inline tests/baselines/reference/typeCheckTypeArgument.js]: Because now we just use the `@noLib` flag

File: tests/baselines/reference/templateStringsArrayTypeRedefinedInES6Mode.errors.txt (hunk 7 above)
Comment (DanielRosenwasser) [inline tests/baselines/reference/templateStringsArrayTypeRedefinedInES6Mode.errors.txt]: Now `TemplateStringsArray` is resolved as the class. Not sure if we should add `@noLib` here to

File: src/harness/harness.ts (hunk 5 above)
Comment (DanielRosenwasser) [inline src/harness/harness.ts]: We were just passing `!Harness.lightMode` everywhere anyway. I can change it back.

File: src/compiler/program.ts [hunk 8]
```diff
@@ -151,18 +151,26 @@ module ts {
         let files: SourceFile[] = [];
         let filesByName: Map<SourceFile> = {};
         let diagnostics = createDiagnosticCollection();
-        let seenNoDefaultLib = options.noLib;
         let commonSourceDirectory: string;
         let diagnosticsProducingTypeChecker: TypeChecker;
         let noDiagnosticsTypeChecker: TypeChecker;
 
+        let encounteredTheNoDefaultLibTag = false;
```
Comment (CyrusNajmabadi) [inline src/compiler/program.ts]: the old name seemed fine.

Comment (CyrusNajmabadi) [inline src/compiler/program.ts]: also, it seems like it could still be initialized to options.noLib

File: src/compiler/program.ts [hunk 9]
```diff
@@ -151,18 +151,26 @@ module ts {
...
-        if (!seenNoDefaultLib) {
-            processRootFile(host.getDefaultLibFileName(options), true);
+
+        forEach(rootNames, name => processRootFile(name, /*isDefaultLib */ false));
+
+        // Do not process the default library if:
+        //  - The '--noLib' flag is used.
+        //  - A 'no-default-lib' reference comment is encountered in
+        //      processing the root files.
+        if (!(options.noLib || encounteredTheNoDefaultLibTag)) {
```
Comment (CyrusNajmabadi) [inline src/compiler/program.ts]: then you wouldn't need to check options.noLib here.

File: tests/baselines/reference/typeCheckTypeArgument.errors.txt [hunk 10]
```diff
@@ -6,12 +6,12 @@ error TS2318: Cannot find global type 'Number'.
 error TS2318: Cannot find global type 'Object'.
 error TS2318: Cannot find global type 'RegExp'.
 error TS2318: Cannot find global type 'String'.
-tests/cases/compiler/typeCheckTypeArgument.ts(3,19): error TS2304: Cannot find name 'UNKNOWN'.
-tests/cases/compiler/typeCheckTypeArgument.ts(5,26): error TS2304: Cannot find name 'UNKNOWN'.
-tests/cases/compiler/typeCheckTypeArgument.ts(7,21): error TS2304: Cannot find name 'UNKNOWN'.
-tests/cases/compiler/typeCheckTypeArgument.ts(9,24): error TS2304: Cannot find name 'UNKNOWN'.
-tests/cases/compiler/typeCheckTypeArgument.ts(12,22): error TS2304: Cannot find name 'UNKNOWN'.
-tests/cases/compiler/typeCheckTypeArgument.ts(15,13): error TS2304: Cannot find name 'UNKNOWN'.
+tests/cases/compiler/typeCheckTypeArgument.ts(2,19): error TS2304: Cannot find name 'UNKNOWN'.
```
Comment (CyrusNajmabadi) [inline tests/baselines/reference/typeCheckTypeArgument.errors.txt]: This seems odd.  All the lines are now off by one?

File: src/harness/harness.ts [hunk 11]
```diff
@@ -732,7 +732,8 @@ module Harness {
     }
 
     // Settings 
-    export var userSpecifiedroot = "";
+    export let userSpecifiedRoot = "";
+    export let lightMode = false;
```
Comment (CyrusNajmabadi) [inline src/harness/harness.ts]: We probably don't want this as a global.  Instead, decide this on a test by test basis.

File: src/harness/harness.ts [hunk 12]
```diff
@@ -959,6 +961,9 @@ module Harness {
                 options.module = options.module || ts.ModuleKind.None;
                 options.newLine = options.newLine || ts.NewLineKind.CarriageReturnLineFeed;
                 options.noErrorTruncation = true;
+                if (lightMode) {
```
Comment (CyrusNajmabadi) [inline src/harness/harness.ts]: have light mode be a local here.  Possibly default it to 'true'.  Then check test settings (which might set it to false if that's what the test does).   Then, from that point on you can use the local.

File: src/harness/harness.ts [hunk 13]
```diff
@@ -1121,7 +1152,7 @@ module Harness {
                         case 'inlinesourcemap':
                             options.inlineSourceMap = setting.value === 'true';
                             break;
-                        
+
```
Comment (CyrusNajmabadi) [inline src/harness/harness.ts]: check for a lightMode setting here.

File: tests/baselines/reference/templateStringsArrayTypeRedefinedInES6Mode.errors.txt (hunk 7 above)
Comment (CyrusNajmabadi) [inline tests/baselines/reference/templateStringsArrayTypeRedefinedInES6Mode.errors.txt]: i thnk you want ```@light: false``` here

File: src/compiler/program.ts (hunk 8 above)
Comment (DanielRosenwasser) [inline src/compiler/program.ts]: Your exact words were

> Merging the two disparate systems into one variable, which you mutate on the fly seems much harder to understand.

**Comment by mhegazy** [comment]
@DanielRosenwasser can you refresh this request..

File: src/harness/harness.ts (hunk 11 above)
Comment (DanielRosenwasser) [inline src/harness/harness.ts]: Since light mode only skips invariant checking now, I don't think this applies anymore.

File: src/harness/harness.ts [hunk 14]
```diff
@@ -827,8 +829,8 @@ module Harness {
         const lineFeed = "\n";
 
         export var defaultLibFileName = 'lib.d.ts';
-        export var defaultLibSourceFile = createSourceFileAndAssertInvariants(defaultLibFileName, IO.readFile(libFolder + 'lib.core.d.ts'), /*languageVersion*/ ts.ScriptTarget.Latest, /*assertInvariants:*/ true);
-        export var defaultES6LibSourceFile = createSourceFileAndAssertInvariants(defaultLibFileName, IO.readFile(libFolder + 'lib.core.es6.d.ts'), /*languageVersion*/ ts.ScriptTarget.Latest, /*assertInvariants:*/ true);
+        export var defaultLibSourceFile = createSourceFileAndAssertInvariants(defaultLibFileName, IO.readFile(libFolder + 'lib.core.d.ts'), /*languageVersion*/ ts.ScriptTarget.Latest, /*assertInvariants*/ !Harness.lightMode);
```
Comment (CyrusNajmabadi) [inline src/harness/harness.ts]: Why not just have createSourceFileAndAssertInvariants check for Harness.lightMode internally?

File: src/harness/harness.ts [hunk 15]
```diff
@@ -939,17 +941,20 @@ module Harness {
...
+                this.compileFiles(this.inputFiles,
+                    /*otherFiles*/ [],
+                    /*onComplete*/ (result) => {
+                        result.files.forEach(writeFile);
+                        result.declFilesCode.forEach(writeFile);
+                        result.sourceMaps.forEach(writeFile);
+                    },
+                    /*settingsCallback*/ () => { },
+                    this.compileOptions,
+                    /*currentDirectory*/ undefined);
```
Comment (CyrusNajmabadi) [inline src/harness/harness.ts]: this is an optional parameter right?  you could leave it off.  ohterwise, make it non-optional.

File: src/compiler/types.ts [hunk 16]
```diff
@@ -1190,7 +1198,7 @@ namespace ts {
         (fileName: string, data: string, writeByteOrderMark: boolean, onError?: (message: string) => void): void;
     }
 
-    export interface Program extends ScriptReferenceHost {
+    export interface Program extends ScriptReferenceHost, TypeCheckerHost {
```
Comment (CyrusNajmabadi) [inline src/compiler/types.ts]: remove.

File: src/harness/harness.ts [hunk 17]
```diff
@@ -939,17 +941,20 @@ module Harness {
...
-                    result.declFilesCode.forEach(file => {
-                        ioHost.writeFile(file.fileName, file.code, false);
-                    });
-                    result.sourceMaps.forEach(file => {
-                        ioHost.writeFile(file.fileName, file.code, false);
-                    });
-                },() => { }, this.compileOptions);
+                this.compileFiles(this.inputFiles,
+                    /*otherFiles*/ [],
+                    /*onComplete*/ (result) => {
```
Comment (CyrusNajmabadi) [inline src/harness/harness.ts]: no parens

File: tests/cases/compiler/noDefaultLib.ts [hunk 18]
```diff
@@ -1,4 +1,4 @@
-/// <reference no-default-lib="true"/>
+// @noLib: true
```
Comment (CyrusNajmabadi) [inline tests/cases/compiler/noDefaultLib.ts]: wut?
